### PR TITLE
collection: filter collaborator emails

### PR DESCRIPTION
**Pull request by diegoviola**
Fixes: #1993

### CODE REVIEW

**Comment by diegoviola** [comment]
This will disable the "All Collaborator Emails" part of the Collaborators tab when toggling on the "Nightly emails with owned collection activity" of a user profile.

**Comment by saracarl** [comment]
This does not accomplish the goal of the issue which is to give transcribers an option to opt-out of collection emails in their user profile, then if they have opted-out remove just their email address from *both* the list of collaborators in that view and in the owner dashboard spreadsheet download.

File: app/controllers/collection_controller.rb [hunk 1]
```diff
@@ -326,37 +326,41 @@ def contributors_download
     ]
 
     stats = @active_transcribers.map do |user|
-      time_total = 0
-      time_proportional = 0
+      if user.notification.user_activity?
```
Comment (saracarl) [inline app/controllers/collection_controller.rb:329]: This doesn't work.  Project owners need to know how much work is happening.  The intent of this fix is just to keep project owners from emailing folks.  We should just skip printing their email address in this report, but include their username and time contributed.

File: app/views/collection/_contributors_body.html.slim [hunk 2]
```diff
@@ -30,7 +31,7 @@
     -unless @active_transcribers.empty?
       p
       -@active_transcribers.each do |user|
-          =link_to user.display_name, user_profile_path(:user_id => user.id, only_path: false)
+          =link_to user.display_name, user_profile_path(:user_id => user.id, only_path: false) if user.notification.user_activity?
```
Comment (saracarl) [inline app/views/collection/_contributors_body.html.slim:34]: Same issue here.  Only skip email address.

File: app/views/collection/_contributors_body.html.slim [hunk 3]
```diff
@@ -49,7 +50,7 @@ h3 =t('.all_collaborator_emails')
 p=link_to "Find recommendations for project newsletters.", "https://content.fromthepage.com/crowdsourcing-project-newsletter/"
 p#collaborators
   -unless @all_collaborators.empty?
-    =@all_collaborators + ", support@fromthepage.com"
+    =format_collaborators
```
Comment (saracarl) [inline app/views/collection/_contributors_body.html.slim:53]: This is the only place that needs support@fromthepage.com.

**Review by saracarl — CHANGES_REQUESTED** [review]

**Comment by benwbrum** [comment]
If user Tom Jones has opted out, he should still appear in the statistics and in the spreadsheet download, but his email should be blank.  His email should also be blank in the contributor email list.

**Comment by benwbrum** [comment]
Users who have opted out should still appear in the statistics and the spreadsheet, but their email should be blank.

File: app/controllers/collection_controller.rb (hunk 1 above)
Comment (diegoviola) [reply app/controllers/collection_controller.rb]: @saracarl OK, I changed the code to show emails as "opt out" for users who have their user_activity set to false. Please let me know if this works.

File: app/helpers/contributor_helper.rb [hunk 4]
```diff
@@ -68,7 +68,20 @@ def new_contributors(collection_id, start_date, end_date)
...
+
+  def format_collaborators
+    @collaborators = @all_collaborators.split
+    last_char = @collaborators.last[-1]
+
+    if last_char == ","
+      @all_collaborators + " support@fromthepage.com"
+    else
+      @all_collaborators + ", support@fromthepage.com"
+    end
```
[the remark this answers is not in the document]
Comment (diegoviola) [reply app/helpers/contributor_helper.rb:84]: @saracarl this is just a helper method used in the view to remove a comma when the comma is already included in the string. It shouldn't affect the CSV at all.

File: app/views/collection/_contributors_body.html.slim [hunk 5]
```diff
@@ -6,10 +6,11 @@
     h3.nomargin =t('.new_collaborators')
     p
       - @new_transcribers.each do |t|
-        =link_to t.display_name, user_profile_path(:user_id => t.id, only_path: false)
-        |
-        =t.email
-        | <br>
+        -if t.notification.user_activity?
```
[the remark this answers is not in the document]
Comment (diegoviola) [reply app/views/collection/_contributors_body.html.slim:11]: @saracarl done.

File: app/views/collection/_contributors_body.html.slim (hunk 3 above)
Comment (diegoviola) [reply app/views/collection/_contributors_body.html.slim:55]: @saracarl right, format_collaborators won't affect the CSV output, it's only there to remove a comma from the view when the string already includes one.

File: app/controllers/collection_controller.rb [hunk 6]
```diff
@@ -337,6 +337,8 @@ def contributors_download
         time_proportional = (@user_time_proportional[user.id] / 60 + 1).floor
       end
 
+      user.email = "opted out" unless user.notification.user_activity?
+
```
Comment (saracarl) [inline app/controllers/collection_controller.rb:341]: Clever.  I like this solution.

**Review by saracarl — CHANGES_REQUESTED** [review]
This is looking good, but there's still 1 piece missing.  Instead of using user.notification.user_activity we need to create a new field user.notification.owner_contact.  It should default to to the user.activity_email (not user.notification.user_activity). 

In the user profile, we should add a "Share my email with project administrators" to set/unset it.

image


And then in this code it should be based on user.notification.owner_contact.

**Review by saracarl — CHANGES_REQUESTED** [review]
Testing this, I get a:
"undefined method `owner_contact?' for nil:NilClass" when I go to the collaborators tab for a collection.  It's in contributor_helper.rb line 72.  The data is an existing user.  The migration defaults to "true", so I'm not sure why this isn't set.

**Pull request closed without merging**